Add _tag and _profile search parameters for x-fhir-query

gosso22 opened the pull request:
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1935

**Description**
In the toSearchParamDefinitionValueMap method, I added _tag and _profile in the paramDefinition.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | Code health | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

maimoonak commented on engine/src/main/java/com/google/android/fhir/search/query/XFhirQueryTranslator.kt:
```diff
@@ -182,7 +182,20 @@ object XFhirQueryTranslator {
-      val paramDefinition = paramKey.toSearchParamDefinition(type)
+      val paramDefinition =
+        when (paramKey) {
+          "_tag" ->
```
We can add _tag and _ profile as Constants?

maimoonak commented on engine/src/test/java/com/google/android/fhir/search/query/XFhirQueryTranslatorTest.kt:
```diff
@@ -337,4 +337,28 @@ class XFhirQueryTranslatorTest {
     assertThat(exception.message).isEqualTo("SPECIAL type not supported in x-fhir-query")
   }
+
+  @Test
```
Add a test in com.google.android.fhir.impl.FhirEngineImplTest#search() by x-fhir-query should return female patients for gender param as well for _tag and _profile for integration testing.

maimoonak commented on engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt:
```diff
@@ -253,6 +255,79 @@ class FhirEngineImplTest {
   }
 
+  @Test
+  fun `search() by x-fhir-query should return patients for gender param as well for _tag`() =
```
Keep it simple. Just run it for tag. Remove gender based search and properties

maimoonak commented on engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt:
```diff
@@ -253,6 +255,79 @@ class FhirEngineImplTest {
+    }
+
+  @Test
+  fun `search() by x-fhir-query should return patients for gender param as well for _profile`() =
```
same as above comment. Keep it simple and just test profile. Remove gender

maimoonak commented on engine/src/main/java/com/google/android/fhir/search/query/XFhirQueryTranslator.kt:
```diff
@@ -182,7 +184,20 @@ object XFhirQueryTranslator {
+      val paramDefinition =
+        when (paramKey) {
+          XFHIR_QUERY_TAG_PARAM ->
+            SearchParamDefinition("_tag", Enumerations.SearchParamType.TOKEN, "$type.meta.tag")
```
use XFHIR_QUERY_TAG_PARAM constant here... do the same for profile

maimoonak commented on engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt:
```diff
@@ -253,6 +255,80 @@ class FhirEngineImplTest {
+                )
+              )
+        },
+        buildPatient("4", "C", Enumerations.AdministrativeGender.MALE).apply {
```
change it FEMALE so that only difference is profile

maimoonak commented on engine/src/main/java/com/google/android/fhir/search/query/XFhirQueryTranslator.kt:
```diff
@@ -182,7 +184,24 @@ object XFhirQueryTranslator {
+              Enumerations.SearchParamType.TOKEN,
+              "$type.meta.tag"
+            )
+          XFHIR_QUERT_PROFILE_PARAM ->
```
Rename spelling to quer**Y**.. its **T** here...

maimoonak reviewed (approved):
Just one issue with spelling.. rest is fine.

gosso22 commented:
@jingtang10 - can you approve and merge these changes?

MJ1998 commented on engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt:
```diff
@@ -253,6 +255,80 @@ class FhirEngineImplTest {
+          meta =
+            Meta()
+              .setTag(
+                mutableListOf(Coding("http://d-tree.org/", "salima-catchment", "Salima Patients"))
```
I think we should have generalised names like Tag1, Patient1

MJ1998 commented on engine/src/main/java/com/google/android/fhir/search/query/XFhirQueryTranslator.kt:
```diff
@@ -47,6 +47,8 @@ import org.hl7.fhir.r4.model.ResourceType
   private const val XFHIR_QUERY_SORT_PARAM = "_sort"
   private const val XFHIR_QUERY_COUNT_PARAM = "_count"
+  private const val XFHIR_QUERY_TAG_PARAM = "_tag"
+  private const val XFHIR_QUERY_PROFILE_PARAM = "_profile"
```
Can we use com.google.android.fhir.index.getSearchParamList instead ?

MJ1998 reviewed (changes requested):
So `_tag` and `_profile` fall under standard search parameters for all resources.
link: https://hl7.org/fhir/search.html#all
For these standard search params, we already have `SearchParamDefinition` defined in `getSearchParamList` under "Resource" option. So instead of rewriting this code we can make advantage of that.
source code: com.google.android.fhir.index.getSearchParamList

gosso22 commented:
> So `_tag` and `_profile` fall under standard search parameters for all resources. link: https://hl7.org/fhir/search.html#all For these standard search params, we already have `SearchParamDefinition` defined in `getSearchParamList` under "Resource" option. So instead of rewriting this code we can make advantage of that. source code: com.google.android.fhir.index.getSearchParamList

@MJ1998 - Thank you for looking at this and for the feedback. XFhirQueryTranslator is already using the method com.google.android.fhir.index.getSearchParamList [here](https://github.com/google/android-fhir/blob/96af7f12cc8f95a2922dd184ef969745a544641a/engine/src/main/java/com/google/android/fhir/search/query/XFhirQueryTranslator.kt#L177), I expected it to pick _tag and _profile however this is not happening and it throws an error that _tag is not found in ResourceName. I am not sure why that is happening, that's why I decided to handle it explicitly.

MJ1998 commented:
> > So `_tag` and `_profile` fall under standard search parameters for all resources. link: https://hl7.org/fhir/search.html#all For these standard search params, we already have `SearchParamDefinition` defined in `getSearchParamList` under "Resource" option. So instead of rewriting this code we can make advantage of that. source code: com.google.android.fhir.index.getSearchParamList
> 
> @MJ1998 - Thank you for looking at this and for the feedback. XFhirQueryTranslator is already using the method com.google.android.fhir.index.getSearchParamList [here](https://github.com/google/android-fhir/blob/96af7f12cc8f95a2922dd184ef969745a544641a/engine/src/main/java/com/google/android/fhir/search/query/XFhirQueryTranslator.kt#L177), I expected it to pick _tag and _profile however this is not happening and it throws an error that _tag is not found in ResourceName. I am not sure why that is happening, that's why I decided to handle it explicitly.

@gosso22 Either we need to modify static method `resourceSearchParameters` so that it searches for standard parameters under "Resource" first and then for the specific resourceType (eg, "Patient") OR we need to modify `SearchParameterRepositoryGenerator.kt` to generate function `getSearchParamList` that also searches for standard parameters under "Resource". I prefer the latter solution.

gosso22 commented:
> > > So `_tag` and `_profile` fall under standard search parameters for all resources. link: https://hl7.org/fhir/search.html#all For these standard search params, we already have `SearchParamDefinition` defined in `getSearchParamList` under "Resource" option. So instead of rewriting this code we can make advantage of that. source code: com.google.android.fhir.index.getSearchParamList
> > 
> > 
> > @MJ1998 - Thank you for looking at this and for the feedback. XFhirQueryTranslator is already using the method com.google.android.fhir.index.getSearchParamList [here](https://github.com/google/android-fhir/blob/96af7f12cc8f95a2922dd184ef969745a544641a/engine/src/main/java/com/google/android/fhir/search/query/XFhirQueryTranslator.kt#L177), I expected it to pick _tag and _profile however this is not happening and it throws an error that _tag is not found in ResourceName. I am not sure why that is happening, that's why I decided to handle it explicitly.
> 
> @gosso22 Either we need to modify static method `resourceSearchParameters` so that it searches for standard parameters under "Resource" first and then for the specific resourceType (eg, "Patient") OR we need to modify `SearchParameterRepositoryGenerator.kt` to generate function `getSearchParamList` that also searches for standard parameters under "Resource". I prefer the latter solution.

@MJ1998 - I am also in favor of the latter solution.

gosso22 commented:
@MJ1998 - I spent sometime to look on how to do the suggested:
_we need to modify SearchParameterRepositoryGenerator.kt to generate function getSearchParamList that also searches for standard parameters under "Resource"._
My thinking is to include the search parameters included in Resource for example _tag, _profile to also be available for other resources in such a way that when I do getSearchParamList for Location, _tag and _profile are also included with their path defined as Location.meta.tag and Location.meta.profile respectively. What do you think? Would this work?

MJ1998 commented:
@gosso22 
Apologies for late response.
Yes your thinking is correct. We are on the same page. But we would like to modify the generation of `getSearchParamList` in a way to search for only a subset of [standard search parameters](https://hl7.org/fhir/search.html#all) for every resource.
`SearchParameterRepository_Generated:getResource` is already a subset of the list given in the above link. 
@jingtang10 Do you think subset defined in `SearchParameterRepository_Generated:getResource` is enough ? More technically, do you think modifying generator to add statement `addAll(getResource())` is the right thing to do ?

MJ1998 commented on engine/src/test/java/com/google/android/fhir/index/SearchParameterRepositoryGeneratedTest.kt:
```diff
@@ -32,22 +32,68 @@ class SearchParameterRepositoryGeneratedTest(private val resource: Resource) {
-        )
-      }
-      .toMutableList()
+    val searchParams = mutableListOf<SearchParamDefinition>()
```
Thanks for modifying this test! 
The current set of supported search parameters for all resources are given in codegen/src/main/res/search-parameters.json. We would somehow like to use that instead of hardcoding the subset here.

gosso22 replied:
For this one, what I found is the resources defined in model classes  inherits from Resource which does not have these standard search parameters annotated `SearchParamDefinition`. I am not sure how to proceed with this one.

MJ1998 replied:
@jingtang10

jingtang10 replied:
i don't think modifying that file would work for this test case. the "expected" search parameters here are generated from using reflection on HAPI structures java classes... so we'd have to hardcode somehow these search parameters that don't exist in those java classes.

i do think @MJ1998 is right in the way that these can be extracted to a separate function just to make it a bit more explicit, something like `fun getBaseSearchParameters(...)`... and add some comments to say that we do this because these don't exist in the HAPI java classes...

MJ1998 commented on buildSrc/src/main/kotlin/codegen/SearchParameterRepositoryGenerator.kt:
```diff
@@ -100,6 +110,19 @@ internal object SearchParameterRepositoryGenerator {
+              name = baseSearchParam.name,
+              paramTypeCode = baseSearchParam.paramTypeCode,
+              path = baseSearchParam.path.replace("Resource", resource)
+            )
```
This is nice and definitely one way of solving! 
But what if instead of adding duplicate codes to every function we just add a new statement like ```"addAll(getResource())"```  ? (Maybe on line 120 of the original file)
We would also **not** need to store baseResourceSearchParams then. Unless storing it makes sense for using it elsewhere.

gosso22 replied:
@MJ1998 - I thought about adding `addAll(getResource())`, but I think the challenge there is the path. For example if we consider tag for  Location resource, with the above statement, the path for tag will be Resource.meta.tag instead of Location.meta.tag. I am no so sure if this works.

MJ1998 replied:
That's correct. Resolving.

jingtang10 commented on buildSrc/src/main/kotlin/codegen/SearchParameterRepositoryGenerator.kt:
optional:

explore the alternative approach of creating a new function that returns the base search parametesr for all resources, and calling that function in `getSearchParamList()` so that you're not adding the same code to each function for each resoure type.

jingtang10 reviewed (changes requested):
i think you might need to change/remove this function `addIndexesFromResourceClass` in `ResourceIndexer.kt` because now the base search parameters are already provided for each resource type when the indexer calls `getSearchParamList(...)` function.

hopefullly you won't need to change the indexer's tests much.

with this change, instead of adding `_tag` and `_profile`, or even `_lastUpdated` etc in the indexer manually, we're  including these base search paramters for each resource type in the search parameter repository. this is a *very good* change - makes our code more unfied.

thanks @gosso22!

gosso22 commented:
> i think you might need to change/remove this function `addIndexesFromResourceClass` in `ResourceIndexer.kt` because now the base search parameters are already provided for each resource type when the indexer calls `getSearchParamList(...)` function.
> 
> hopefullly you won't need to change the indexer's tests much.
> 
> with this change, instead of adding `_tag` and `_profile`, or even `_lastUpdated` etc in the indexer manually, we're including these base search paramters for each resource type in the search parameter repository. this is a **very good** change - makes our code more unfied.
> 
> thanks @gosso22!

@jingtang10 - I removed the above function `addIndexesFromResourceClass` in the `ResourceIndexer.kt`, and change a few things. First I changed the type for _profile in `search-parameters.json` from uri to reference as defined in the fhir specification and also as it was implemented in `addIndexesFromResourceClass`. Also I had to handle the case for _id. But other than that all tests passed without the `addIndexesFromResourceClass`.

gosso22 commented:
@jingtang10, @aditya-07 - I have updated this, I think it is ready.

aditya-07 commented on buildSrc/src/main/resources/search-parameters.json:
```diff
@@ -189,7 +189,7 @@
       "code" : "_profile",
       "base" : ["Resource"],
-      "type" : "uri",
+      "type" : "reference",
```
I am not sure about the rationale behind **_profile** being added as a reference in [here](https://github.com/google/android-fhir/blob/f05cd76e84b91643313093bc394128f1303efcdc/engine/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt#L144). But, I think we should keep the search-parameters.json as-is since its dowloaded directly from the [registry](https://www.hl7.org/fhir/searchparameter-registry.html).

gosso22 replied:
@jingtang10 and @aditya-07 - I have resolved this and modified the tests that were using reference.

jingtang10 reviewed (commented):
Thanks for this change! I love this PR!

Once @aditya-07 's comment is addressed in the json file i'm happy to approve this!

gosso22 commented on buildSrc/src/main/kotlin/codegen/SearchParameterRepositoryGenerator.kt:
```diff
@@ -41,8 +42,10 @@ import org.hl7.fhir.r4.model.SearchParameter
- * the content at `http://www.hl7.org/fhir/search-parameters.json` and run the `main` function in
- * the `codegen` module.
+ * the content at `http://www.hl7.org/fhir/search-parameters.json` and execute the gradle task
+ * `generateSearchParamsTask`. If you are using Android Studio, you can usually find this task in
```
Thanks @jingtang10
